tests/integration/test_templates.py

```diff
@@ -0,0 +1,126 @@
+    aconfig = {
+        "global": {"http_config": {"tls_config": {"insecure_skip_verify": True}}},
+        "route": {
+            "group_by": ["alertname"],
+            "group_wait": "3s",
+            "group_interval": "5m",
+            "repeat_interval": "1h",
+            "receiver": receiver_name,
+        },
+        "receivers": [
+            {
+                "name": receiver_name,
+                "slack_configs": [
+                    {
+                        "api_url": httpserver.url_for("/"),
+                        "channel": "test",
+                        "text": r"https://localhost/alerts/{{ .GroupLabels.alertname }}",
+                    }
+                ],
+            }
+        ],
+    }
+
+    # use a template to define the slack callback id
+    atemplate = r'{{ define "slack.default.callbackid" }}2{{ end }}'
+    # set alertmanager configuration and template file
+    await ops_test.model.applications[app_name].set_config(
+        {"config_file": yaml.safe_dump(aconfig), "templates_file": atemplate}
+    )
+    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=60)
+
+    # create an alert
+    start_time = datetime.now(timezone.utc)
+    end_time = start_time + timedelta(minutes=5)
+    alert_name = "fake-alert"
+    model_uuid = "1234"
+    alerts = [
+        {
+            "startsAt": start_time.isoformat("T"),
+            "endsAt": end_time.isoformat("T"),
+            "status": "firing",
+            "annotations": {
+                "summary": "A fake alert",
+            },
+            "labels": {
+                "juju_model_uuid": model_uuid,
+                "juju_application": app_name,
+                "juju_model": ops_test.model_name,
+                "alertname": alert_name,
+            },
+            "generatorURL": f"http://localhost/{alert_name}",
+        }
+    ]
+
+    # define the expected slack notification for the alert
+    expected_notification = {
+        "channel": "test",
+        "username": "Alertmanager",
+        "attachments": [
+            {
+                "title": f"[FIRING:1] {alert_name} {app_name} {ops_test.model_name} {model_uuid} ",
+                "title_link": f"http://{app_name}-0:9093/#/alerts?receiver={receiver_name}",
+                "text": f"https://localhost/alerts/{alert_name}",
+                "fallback": f"[FIRING:1] {alert_name} {app_name} {ops_test.model_name} {model_uuid}  | "
+                f"http://{app_name}-0:9093/#/alerts?receiver={receiver_name}",
+                "callback_id": "2",
+                "footer": "",
+                "color": "danger",
+                "mrkdwn_in": ["fallback", "pretext", "text"],
+            }
+        ],
+    }
+
+    # set the alert
+    with httpserver.wait(timeout=120) as waiting:
+        # expect an alert to be forwarded to the receiver
+        httpserver.expect_oneshot_request(
+            "/", method="POST", json=expected_notification
```

Comment:
Passing `json=expected_notification` seems fragile.
Is there a way to simply assert that e.g. `"callback_id": "2"` is in the posted request?

Reply:
No simple way that I am aware of.